### DISCUSSION TITLE
asp.net core (web application) support

### DISCUSSION THREAD

**Discussion by pkrukp** [OpenCover/opencover#1023] [Q&A] [answered]
Hi,

is collecting coverage from running **Web Application (asp.net core 3.x, 5.x)** supported? I tried **OpenCover 4.7.1189** but wasn't able to do it.

I saw mentions that .net core is supported (e.g. https://github.com/OpenCover/opencover/issues/595), OTOH https://github.com/OpenCover/opencover/discussions/1016 says "from the next release" so it's a bit confusing...

I created a simple demo web application, for example: `dotnet new mvc -f netcoreapp3.1` or for .net 5.0

Then I tried collecting coverage using various opencover command lines, did not succeed.

For example:

`OpenCover.Console.exe "-target:C:\Program Files\dotnet\dotnet.exe" -targetargs:run -register:user -oldstyle`
`OpenCover.Console.exe "-target:./bin/Debug/net5.0/WebAppNet5.exe"  -register:user -oldstyle`
Similar without `-oldstyle` and with `-register`

Web application starts, I click on it in web browser, then close it in console with Ctrl-C. 
results.xml is empty.

`OpenCover.Console.exe "-target:C:\Program Files\dotnet\dotnet.exe" -targetargs:run -register:user -oldStyle  -targetdir:bin/Debug/net5.0`

Here web application does not seem to start. 
results.xml is not empty, but it only lists several system assemblies as skipped. 
Console output is following:

```Executing: C:\Program Files\dotnet\dotnet.exe
Couldn't find a project to run. Ensure a project exists in C:\Users\pkruk\source\repos\WebAppNet5\bin\Debug\net5.0, or pass the path to the project using --project.
Committing...
No results, this could be for a number of reasons. The most common reasons are:
    1) missing PDBs for the assemblies that match the filter please review the
    output file and refer to the Usage guide (Usage.rtf) about filters.
    2) the profiler may not be registered correctly, please refer to the Usage
    guide and the -register switch.
    3) your assemblies under test were not be loaded, refer to the output xml
    and check your filters/code.
    4) you are targetting .net core and your assemblies under test were not
    instrumented, try using the -oldstyle switch.
    5) you are targetting 4.8 or .net core and your assemblies under test were
    not instrumented, try using the -register:path32 or -register:path64 switch.
```

**Comment by sawilde** [comment]
I've not tried. OpenCover is a code coverage tool that is really meant for use against testing frameworks such as NUnit, XUnit, VSTest and their runners and this is how I use it.

If you have a sample that repeats the issue I can try then I may be able to help diagnose the issue if I have the time.

**Reply by pkrukp** [reply]
Yes. I'm attaching simple apps. You can also create them by yourself like this:
```
dotnet new console -o console5
dotnet new mvc -o webapp5
```
[console5.zip](https://github.com/OpenCover/opencover/files/6494067/console5.zip)
[webapp5.zip](https://github.com/OpenCover/opencover/files/6494068/webapp5.zip)

Then you can compile and make sure you can run them, and check if opencover collects coverage:

```
$ cd console5
$ dotnet run
Hello World!
$ OpenCover.Console.exe "-target:C:\Program Files\dotnet\dotnet.exe" -targetargs:run -register:user
Executing: C:\Program Files\dotnet\dotnet.exe
Hello World!
Committing...
Visited Classes 1 of 1 (100)
Visited Methods 1 of 1 (100)
Visited Points 3 of 3 (100)
Visited Branches 1 of 1 (100)

==== Alternative Results (includes all methods including those without corresponding source) ====
Alternative Visited Classes 1 of 1 (100)
Alternative Visited Methods 1 of 2 (50)
```

For web application I have passed `--no-build` to `dotnet run`, otherwise it tries to compile code.
```
$ cd ../webapp5
$ dotnet build
Microsoft (R) Build Engine version 16.9.0+57a23d249 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
  webapp5 -> C:\Users\pkruk\source\repos\webapp5\bin\Debug\net5.0\webapp5.dll
  webapp5 -> C:\Users\pkruk\source\repos\webapp5\bin\Debug\net5.0\webapp5.Views.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:00.92

$ dotnet run --no-build
info: Microsoft.Hosting.Lifetime[0]
      Now listening on: https://localhost:5001
info: Microsoft.Hosting.Lifetime[0]
      Now listening on: http://localhost:5000
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Development
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\Users\pkruk\source\repos\webapp5
info: Microsoft.Hosting.Lifetime[0]
      Application is shutting down...

$ OpenCover.Console.exe "-target:C:\Program Files\dotnet\dotnet.exe" "-targetargs:run --no-build" -register:user
Executing: C:\Program Files\dotnet\dotnet.exe
info: Microsoft.Hosting.Lifetime[0]
      Now listening on: https://localhost:5001
info: Microsoft.Hosting.Lifetime[0]
      Now listening on: http://localhost:5000
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Development
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\Users\pkruk\source\repos\webapp5
info: Microsoft.Hosting.Lifetime[0]
      Application is shutting down...
```

**Reply by sawilde** [reply]
The profiler is hooking-in ([debugview](https://docs.microsoft.com/en-us/sysinternals/downloads/debugview) logging) but it looks like the "ctrl+c" used to kill the `dotnet` process is also killing the `opencover.console` host process and so no results are returned.

We should be able to intercept this

**Reply by sawilde** [reply]
Try this build https://github.com/OpenCover/opencover/pull/1028 - added a `-ignorectrlc` command line option.

Note: Use with care.

**Reply by pkrukp** [reply]
Thanks, it's working! :)

BTW what do you mean by "use with care" ?

**Reply by sawilde** [reply]
thanks 

> BTW what do you mean by "use with care" ?

as in it has only been tested with this scenario in mind and you are the first person to use it :)

**Comment by pkrukp** [comment, marked as answer]
Build #1028 solved it

(Posting new comment so I can mark this as "answered" - seems you can't mark replies as "answers")
